prometheus/client.go

```diff
@@ -0,0 +1,352 @@
+	}
+}
+
+func buildStatusRequest(prometheusURL string) (*http.Request, error) {
+	finalURL, err := url.Parse(prometheusURL)
+	if err != nil {
+		return nil, err
+	}
+	finalURL.Path = "/api/v1/status/buildinfo"
+	httpRequest, err := http.NewRequest(http.MethodGet, finalURL.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	// set the accept content type
+	httpRequest.Header.Set("Accept", "application/json")
+	return httpRequest, nil
+}
+
+type buildInfoResponse struct {
+	Status    string        `json:"status"`
+	Data      buildInfoData `json:"data,omitempty"`
+	ErrorType string        `json:"errorType,omitempty"`
+	Error     string        `json:"error,omitempty"`
+	Warnings  []string      `json:"warnings,omitempty"`
+}
+
+// buildInfoData contains build information about Prometheus.
+type buildInfoData struct {
+	Version   string `json:"version"`
+	Revision  string `json:"revision"`
+	Branch    string `json:"branch"`
+	BuildUser string `json:"buildUser"`
+	BuildDate string `json:"buildDate"`
+	GoVersion string `json:"goVersion"`
+}
+
+type Client interface {
+	Metadata(metric string) (v1.Metadata, error)
+	AllMetadata() (map[string][]v1.Metadata, error)
+	LabelNames(metricName string) ([]string, error)
+	LabelValues(label string) ([]model.LabelValue, error)
+	ChangeDataSource(prometheusURL string) error
+	// GetURL is returning the url used to contact the prometheus server
+	// In case the instance is used directly in Prometheus, it should be the externalURL
+	GetURL() string
+}
+
+// httpClient is an implementation of the interface Client.
+// You should use this instance directly and not the other one (compatibleHTTPClient and notCompatibleHTTPClient)
+// because it will manage which sub instance of the Client to use (like a factory)
+type httpClient struct {
+	Client
+	requestTimeout time.Duration
+	mutex          sync.RWMutex
+	subClient      Client
+	url            string
+}
+
+func NewClient(prometheusURL string) (Client, error) {
+	c := &httpClient{
+		requestTimeout: 30 * time.Second,
+	}
+	if err := c.ChangeDataSource(prometheusURL); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (c *httpClient) Metadata(metric string) (v1.Metadata, error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.subClient.Metadata(metric)
+}
+
+func (c *httpClient) AllMetadata() (map[string][]v1.Metadata, error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.subClient.AllMetadata()
+}
+
+func (c *httpClient) LabelNames(name string) ([]string, error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.subClient.LabelNames(name)
+}
+
+func (c *httpClient) LabelValues(label string) ([]model.LabelValue, error) {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.subClient.LabelValues(label)
+}
+
+func (c *httpClient) GetURL() string {
+	c.mutex.RLock()
+	defer c.mutex.RUnlock()
+	return c.url
+}
+
+func (c *httpClient) ChangeDataSource(prometheusURL string) error {
```

Comment:
If something goes wrong here or the URL passed is an empty string (e.g. when the user of the VS Code extension removes the Prometheus URL from the respective settings field), we should end up with a `httpClient` instance that can still be used, but always returns empty results when queried.

Reply:
Mmm ok, but in any case does it still return an error ? Because if an empty URL is valid, maybe in that case we shouldn't return an error saying the url is not valid ? (That's what `url.Parse()` will return)

Reply:
I finally have chosen this direction. I'm filling it's simpler like that.
Tell me if you have a different opinion on that.

Reply:
The current approach looks ok to me.

But when I'm trying to connect to an actual Prometheus (running on `localhost`), as described [here](https://github.com/prometheus-community/promql-langserver/blob/master/CONTRIBUTING.md#manually-testing-the-language-server) and [here](https://github.com/redhat-developer/vscode-promql#trying-out-the-extension), I get an connection timeout:
```
Failed to connect to Prometheus at http://localhost:9090: Get "http://localhost:9090/api/v1/status/buildinfo": dial tcp: i/o timeout
```

Reply:
That's weird. I certainly broke something during the changes :/

Reply:
ok my bad, I broke the timeout parameter. Should be ok now